lib/compress/zstd_opt.c

```diff
@@ -423,237 +424,241 @@ size_t ZSTD_compressBlock_opt_generic(ZSTD_CCtx* ctx,
+                best_off = matches[nb_matches-1].off;
+                cur = 0;
+                last_pos = 1;
+                goto _shortestPath;
+            }
 
-        /* set prices using matches at position = 0 */
-        best_mlen = (last_pos) ? last_pos : minMatch;
-        for (u = 0; u < match_num; u++) {
-            mlen = (u>0) ? matches[u-1].len+1 : best_mlen;
-            best_mlen = matches[u].len;
-            while (mlen <= best_mlen) {
-                price = ZSTD_getPrice(optStatePtr, litlen, anchor, matches[u].off-1, mlen - MINMATCH, ultra);
-                if (mlen > last_pos || price < opt[mlen].price)
-                    SET_PRICE(mlen, mlen, matches[u].off, litlen, price);   /* note : macro modifies last_pos */
-                mlen++;
-        }   }
+            /* set prices for first matches from position == 0 */
+            {   U32 matchNb;
+                U32 pos = last_pos /*some repCode (assumed cheaper)*/ ? last_pos : minMatch;
+                for (matchNb = 0; matchNb < nb_matches; matchNb++) {
+                    U32 const end = matches[matchNb].len;
+                    while (pos <= end) {
+                        U32 const matchPrice = ZSTD_getPrice(optStatePtr, initLL, anchor, matches[matchNb].off-1, pos, ultra);
+                        if (pos > last_pos || matchPrice < opt[pos].price)
+                            SET_PRICE(pos, pos, matches[matchNb].off, initLL, matchPrice);   /* note : macro modifies last_pos */
+                        pos++;
+        }   }   }   }
 
         if (last_pos < minMatch) { ip++; continue; }
 
         /* initialize opt[0] */
         { U32 i ; for (i=0; i<ZSTD_REP_NUM; i++) opt[0].rep[i] = rep[i]; }
         opt[0].mlen = 1;
-        opt[0].litlen = litlen;
+        opt[0].litlen = initLL;
 
-         /* check further positions */
+        /* check further positions */
         for (cur = 1; cur <= last_pos; cur++) {
-           inr = ip + cur;
+            const BYTE* const inr = ip + cur;
+            assert(cur < ZSTD_OPT_NUM);
 
-           if (opt[cur-1].mlen == 1) {
-                litlen = opt[cur-1].litlen + 1;
+            /* Fix current position with one literal if cheaper */
+            {   U32 const litlen = (opt[cur-1].mlen == 1) ? opt[cur-1].litlen + 1 : 1;
+                U32 price;
                 if (cur > litlen) {
                     price = opt[cur - litlen].price + ZSTD_getLiteralPrice(optStatePtr, litlen, inr-litlen);
-                } else
+                } else {
                     price = ZSTD_getLiteralPrice(optStatePtr, litlen, anchor);
-           } else {
-                litlen = 1;
-                price = opt[cur - 1].price + ZSTD_getLiteralPrice(optStatePtr, litlen, inr-1);
-           }
-
-           if (cur > last_pos || price <= opt[cur].price)
-                SET_PRICE(cur, 1, 0, litlen, price);
+                }
+                if (cur > last_pos || price <= opt[cur].price)
```

Comment:
All the previous checks have been `price < opt[cur].price`, is there a reason for this to be different? If so can you add a comment.

Reply:
Good point !
Not sure about this one.
It's not a normal match here, it's a literal replacement.
So this `<=` favor literals over matches.

Not sure of the impact.
First, equality is less common in `zstd` compared to `lz4`, as price is expressed in bits, instead of bytes.
Second, all costs are "evaluated", which basically means they are wrong by a small amount. I noticed in the code many heuristics that try to compensate for that in bizarre ways. That could be one of them.

A simple answer to this `<` vs `<=` is : try both, keep the better one.